In [5]:
#Q1
import pandas as pd
df = pd.read_csv("C:/Users/TANISHKA/OneDrive/Desktop/ML and GenAI Internship/Assignment/Week 2/Dataset 2.csv")
df.head()



,UserID,Age,Gender,SubscriptionType,WatchHoursPerWeek,DevicesUsed,FavoriteGenre,AdClicks,MonthlySpend,SubscriptionRenewed
0,1001,22,Female,Basic,23,5,Comedy,13,353,No
1,1002,55,Male,Basic,9,4,Drama,14,317,Yes
2,1003,49,Male,Basic,8,3,Comedy,16,309,No
3,1004,39,Female,Premium,19,5,Drama,45,833,Yes
4,1005,38,Female,Premium,23,5,Sci-Fi,24,804,Yes


In [6]:
#Q2
print("Number of Rows:", df.shape[0])
print("Number of Columns:", df.shape[1])

Number of Rows: 750
Number of Columns: 10


In [7]:
#Q3
print(list(df.columns))

['UserID', 'Age', 'Gender', 'SubscriptionType', 'WatchHoursPerWeek', 'DevicesUsed', 'FavoriteGenre', 'AdClicks', 'MonthlySpend', 'SubscriptionRenewed']


In [8]:
#Q4
n_features = df.select_dtypes(include=['int64', 'float64']).columns

c_features = df.select_dtypes(include=['object']).columns

print("Numerical Features:")
print(n_features)

print("\nCategorical Features:")
print(c_features)

Numerical Features:
Index(['UserID', 'Age', 'WatchHoursPerWeek', 'DevicesUsed', 'AdClicks',
       'MonthlySpend'],
      dtype='str')

Categorical Features:
Index(['Gender', 'SubscriptionType', 'FavoriteGenre', 'SubscriptionRenewed'], dtype='str')


C:\Users\TANISHKA\AppData\Local\Temp\ipykernel_14684\1373504227.py:4: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  c_features = df.select_dtypes(include=['object']).columns


In [9]:
#Q5
print("Total Missing Values:", df.isnull().sum().sum())

Total Missing Values: 0


In [10]:
#Q6
average_age = df['Age'].mean()
print("Average Age:", average_age)

Average Age: 41.824


In [11]:
#Q7
average_watch_hours = df['WatchHoursPerWeek'].mean()
print("Average Watch Hours Per Week:", average_watch_hours)

Average Watch Hours Per Week: 14.236


In [12]:
#Q8
average_spending = df['MonthlySpend'].mean()
print("Average Monthly Spending:", average_spending)

Average Monthly Spending: 689.9053333333334


In [13]:
#Q9
subscription_count = df['SubscriptionType'].value_counts()
print(subscription_count)

SubscriptionType
Basic      342
Premium    279
VIP        129
Name: count, dtype: int64


In [14]:
#Q10
renewed_percentage = (
    (df['SubscriptionRenewed'] == 'Yes').mean()
) * 100
print("Renewal Percentage:", renewed_percentage)

Renewal Percentage: 46.266666666666666


In [18]:
#Q11
print(df.select_dtypes(include='object').columns)

from sklearn.preprocessing import LabelEncoder
df_encoded = df.copy()
le = LabelEncoder()
df_encoded['Gender'] = le.fit_transform(df_encoded['Gender'])
df_encoded['SubscriptionType'] = le.fit_transform(df_encoded['SubscriptionType'])
df_encoded['FavoriteGenre'] = le.fit_transform(df_encoded['FavoriteGenre'])
df_encoded['SubscriptionRenewed'] = le.fit_transform(df_encoded['SubscriptionRenewed'])
df_encoded.head()

print(df_encoded.dtypes)

C:\Users\TANISHKA\AppData\Local\Temp\ipykernel_14684\3450868902.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  print(df.select_dtypes(include='object').columns)


Index(['Gender', 'SubscriptionType', 'FavoriteGenre', 'SubscriptionRenewed'], dtype='str')
UserID                 int64
Age                    int64
Gender                 int64
SubscriptionType       int64
WatchHoursPerWeek      int64
DevicesUsed            int64
FavoriteGenre          int64
AdClicks               int64
MonthlySpend           int64
SubscriptionRenewed    int64
dtype: object


In [19]:
#Q12
X = df_encoded.drop(['UserID', 'SubscriptionRenewed'], axis=1)

y = df_encoded['SubscriptionRenewed']

print("Features (X):")
print(X.head())

print("\nTarget (y):")
print(y.head())

print("Shape of X:", X.shape)
print("Shape of y:", y.shape)

Features (X):
   Age  Gender  SubscriptionType  WatchHoursPerWeek  DevicesUsed  \
0   22       0                 0                 23            5   
1   55       1                 0                  9            4   
2   49       1                 0                  8            3   
3   39       0                 1                 19            5   
4   38       0                 1                 23            5   

   FavoriteGenre  AdClicks  MonthlySpend  
0              1        13           353  
1              2        14           317  
2              1        16           309  
3              2        45           833  
4              5        24           804  

Target (y):
0    0
1    1
2    0
3    1
4    1
Name: SubscriptionRenewed, dtype: int64
Shape of X: (750, 8)
Shape of y: (750,)


In [20]:
#Q13
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42
)

print("X_train shape:", X_train.shape)
print("X_test shape :", X_test.shape)

print("y_train shape:", y_train.shape)
print("y_test shape :", y_test.shape)

X_train shape: (600, 8)
X_test shape : (150, 8)
y_train shape: (600,)
y_test shape : (150,)


In [21]:
#Q14
from sklearn.tree import DecisionTreeClassifier

dt_model = DecisionTreeClassifier(random_state=42)
dt_model.fit(X_train, y_train)

y_pred_dt = dt_model.predict(X_test)
print(y_pred_dt[:10])

[0 0 0 1 0 0 0 1 1 1]


In [22]:
#Q15
from sklearn.metrics import accuracy_score

dt_accuracy = accuracy_score(y_test, y_pred_dt)
print("Decision Tree Accuracy:", dt_accuracy)

print("Accuracy Percentage:", dt_accuracy * 100)

Decision Tree Accuracy: 0.56
Accuracy Percentage: 56.00000000000001


In [23]:
#Q16
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(y_test, y_pred_dt)
print(cm)

[[48 34]
 [32 36]]


In [24]:
#Q17
from sklearn.neighbors import KNeighborsClassifier

knn_model = KNeighborsClassifier(n_neighbors=5)
knn_model.fit(X_train, y_train)

y_pred_knn = knn_model.predict(X_test)
print(y_pred_knn[:10])

[1 1 1 1 1 1 1 1 1 1]


In [25]:
#Q18
from sklearn.metrics import accuracy_score

knn_accuracy = accuracy_score(y_test, y_pred_knn)
print("KNN Accuracy:", knn_accuracy)
print("KNN Accuracy Percentage:", knn_accuracy * 100)

print("Decision Tree Accuracy:", dt_accuracy)
print("KNN Accuracy:", knn_accuracy)

if knn_accuracy > dt_accuracy:
    print("KNN performed better than Decision Tree.")
elif knn_accuracy < dt_accuracy:
    print("Decision Tree performed better than KNN.")
else:
    print("Both models achieved the same accuracy.")

KNN Accuracy: 0.6266666666666667
KNN Accuracy Percentage: 62.66666666666667
Decision Tree Accuracy: 0.56
KNN Accuracy: 0.6266666666666667
KNN performed better than Decision Tree.


In [26]:
#Q19
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

X_reg = df_encoded.drop(['UserID', 'MonthlySpend'], axis=1)
y_reg = df_encoded['MonthlySpend']

X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(
    X_reg,
    y_reg,
    test_size=0.20,
    random_state=42
)

lr_model = LinearRegression()
lr_model.fit(X_train_reg, y_train_reg)

r2_score = lr_model.score(X_test_reg, y_test_reg)
print("R² Score:", r2_score)

R² Score: 0.966526665866529


In [27]:
#Q20
print(df_encoded.head())

new_user = pd.DataFrame({
    'Age': [30],
    'Gender': [1],
    'SubscriptionType': [2],
    'WatchHoursPerWeek': [20],
    'DevicesUsed': [3],
    'FavoriteGenre': [1],
    'AdClicks': [10],
    'SubscriptionRenewed': [1]
})
predicted_spend = lr_model.predict(new_user)
print("Predicted Monthly Spending: ₹", predicted_spend[0])

   UserID  Age  Gender  SubscriptionType  WatchHoursPerWeek  DevicesUsed  \
0    1001   22       0                 0                 23            5   
1    1002   55       1                 0                  9            4   
2    1003   49       1                 0                  8            3   
3    1004   39       0                 1                 19            5   
4    1005   38       0                 1                 23            5   

   FavoriteGenre  AdClicks  MonthlySpend  SubscriptionRenewed  
0              1        13           353                    0  
1              2        14           317                    1  
2              1        16           309                    0  
3              2        45           833                    1  
4              5        24           804                    1  
Predicted Monthly Spending: ₹ 1340.7955241071495


# Business Reflection Questions

Q1. Which factors influence subscription renewal the most?

WatchHoursPerWeek

MonthlySpend

SubscriptionType

DevicesUsed

Q2. Why is subscription renewal a classification problem?

Subscription renewal has only two outcomes: Yes or No. Since the target variable belongs to predefined categories, it is a classification problem.

Q3. Why is monthly spending a regression problem?

Monthly spending is a continuous numerical value that can take many possible values. Therefore, predicting it is a regression problem.

Q4. Which algorithm performed better?

Decision Tree Accuracy = 56.00000000000001%

KNN Accuracy = 62.66666666666667%

KNN performed better because it achieved higher accuracy.

Q5. How could Netflix use these predictions?

Identify customers likely to leave

Offer discounts

Send personalized recommendations

Improve customer retention

Increase revenue 
